couchdb/source/pkg/adapter/adapter.go

```diff
@@ -18,68 +18,87 @@ package adapter
 
 import (
 	"context"
+	"io/ioutil"
 	"time"
 
 	cloudevents "github.com/cloudevents/sdk-go"
 	"go.uber.org/zap"
 	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/eventing/pkg/adapter"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/source"
 
 	"knative.dev/eventing-contrib/couchdb/source/pkg/apis/sources/v1alpha1"
 	"knative.dev/eventing-contrib/couchdb/source/pkg/couchdb"
 )
 
-// Hold options for the adapter
-type Options struct {
-	EventSource string
-	CouchDbURL  string
-	Username    string
-	Password    string
-	Database    string
-	Namespace   string
-}
+type envConfig struct {
+	adapter.EnvConfig
 
-// Adapter converts incoming CouchDb events to CloudEvents
-type Adapter interface {
-	Start(stopCh <-chan struct{}) error
+	CouchDbCredentialsPath string `envconfig:"COUCHDB_CREDENTIALS" required:"true"`
+	Database               string `envconfig:"COUCHDB_DATABASE" required:"true"`
+	EventSource            string `envconfig:"EVENT_SOURCE" required:"true"`
 }
 
-type adapter struct {
+type couchDbAdapter struct {
+	namespace string
+	ce        cloudevents.Client
+	reporter  source.StatsReporter
+	logger    *zap.SugaredLogger
+
 	source     string
-	ce         cloudevents.Client
-	namespace  string
-	logger     *zap.SugaredLogger
 	couchDB    couchdb.Connection
 	database   string
 	changeArgs couchdb.ChangeArguments
 }
 
-// New creates an adapter to convert incoming CouchDb changes events to CloudEvents and
-// then sends them to the specified Sink
-func New(ceClient cloudevents.Client, logger *zap.SugaredLogger, options *Options) (Adapter, error) {
-	connection, err := couchdb.Connect(options.CouchDbURL, options.Username, options.Password, time.Minute)
+func NewEnvConfig() adapter.EnvConfigAccessor {
+	return &envConfig{}
+}
+
+func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client, reporter source.StatsReporter) adapter.Adapter {
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	rawurl, err := ioutil.ReadFile(env.CouchDbCredentialsPath + "/url")
 	if err != nil {
-		return nil, err
+		logger.Fatal("Missing url key in secret", zap.Error(err))
+	}
+
+	rawusername, err := ioutil.ReadFile(env.CouchDbCredentialsPath + "/username")
+	if err != nil {
+		logger.Fatal("Missing username key in secret", zap.Error(err))
 	}
 
-	a := &adapter{
-		ce:         ceClient,
-		logger:     logger,
-		namespace:  options.Namespace,
+	rawpassword, err := ioutil.ReadFile(env.CouchDbCredentialsPath + "/password")
```

Comment:
Not for now, but open issue to support secrets ?

Reply:
Kafka has that - and we could see to share some ?

Reply:
the adapter deployment mounts the secret.